R/combineResults.R

```diff
@@ -0,0 +1,232 @@
+#' \strong{Option 3} (the method that this function uses) would be to perform classification separately within each reference, then collate
+#' the results to choose the label with the highest score across references. 
+#' This avoids the potential for reference-specific markers (like option 1) and the need for explicit harmonization (like option 2).
+#' This option leaves a mixture of labels in the final results that is up to the user to resolve.
+#' 
+#' @author Jared Andrews
+#'
+#' @examples
+#' ##############################
+#' ## Mocking up training data ##
+#' ##############################
+#'
+#' Ngroups <- 5
+#' Ngenes <- 1000
+#' means <- matrix(rnorm(Ngenes*Ngroups), nrow=Ngenes)
+#' means[1:900,] <- 0
+#' colnames(means) <- LETTERS[1:5]
+#'
+#' g <- rep(LETTERS[1:5], each=4)
+#' g2 <- rep(LETTERS[6:10], each=4)
+#' ref1 <- SummarizedExperiment(
+#'     list(counts=matrix(rpois(1000*length(g), 
+#'     lambda=10*2^means[,g]), ncol=length(g))),
+#'     colData=DataFrame(label=g)
+#' )
+#' ref2 <- SummarizedExperiment(
+#'     list(counts=matrix(rpois(1000*length(g2), 
+#'     lambda=10*2^means[,g]), ncol=length(g2))),
+#'     colData=DataFrame(label=g2)
+#' )
+#' rownames(ref1) <- sprintf("GENE_%s", seq_len(nrow(ref1)))
+#' rownames(ref2) <- sprintf("GENE_%s", seq_len(nrow(ref2)))
+#'
+#' ref1 <- scater::logNormCounts(ref1)
+#' ref2 <- scater::logNormCounts(ref2)
+#' trained1 <- trainSingleR(ref1, ref1$label)
+#' trained2 <- trainSingleR(ref2, ref2$label)
+#'
+#' ###############################
+#' ## Mocking up some test data ##
+#' ###############################
+#'
+#' N <- 100
+#' g <- sample(LETTERS[1:5], N, replace=TRUE)
+#' means <- matrix(rnorm(Ngenes*Ngroups), nrow=Ngenes)
+#' means[1:900] <- 0
+#' colnames(means) <- LETTERS[1:5]
+#' test <- SummarizedExperiment(
+#'     list(counts=matrix(rpois(1000*N, lambda=2^means[,g]), ncol=N)),
+#'     colData=DataFrame(label=g)
+#' )
+#' 
+#' rownames(test) <- sprintf("GENE_%s", seq_len(nrow(test)))
+#' test <- scater::logNormCounts(test)
+#' 
+#' ###############################
+#' ## Performing classification ##
+#' ###############################
+#' 
+#' pred1 <- SingleR(test, ref1, labels=ref1$label)
+#' pred2 <- SingleR(test, ref2, labels=ref2$label)
+#'
+#' pred3 <- SingleR(test, ref1, labels=ref1$label, 
+#'     method="cluster", clusters=test$label) 
+#' pred4 <- SingleR(test, ref2, labels=ref2$label, 
+#'     method="cluster", clusters=test$label) 
+#'
+#' ###############################
+#' ##     Combining results     ##
+#' ###############################
+#'
+#' pred.single <- combineResults(list(pred1, pred2))
+#' pred.clust <- combineResults(list(pred3, pred4))
+#'
+#' @seealso
+#' \code{\link{matchReferences}}, to harmonize labels between reference datasets.
+#' \code{\link{SingleR}}, for generating predictions.
+#'
+#' @importFrom S4Vectors DataFrame metadata metadata<-
+#'
+#' @export
+combineResults <- function(results) {
+    num.features <- sapply(results, nrow)
```

Comment:
Avoid `sapply` due to the difficulty in predicting its behavior, use `vapply(results, nrow, 0L)` instead.